guava/src/com/google/common/io/ByteStreams.java

```diff
@@ -899,14 +914,16 @@ private static long skipSafely(InputStream in, long n) throws IOException {
   @CanIgnoreReturnValue
   // Sometimes you don't care how many bytes you actually read, I guess.
   // (You know that it's either going to read len bytes or stop at EOF.)
-  public static int read(InputStream in, byte[] b, int off, int len) throws IOException {
+  public static int read(InputStream in, byte[] b, @IndexOrHigh("#2") int off, @NonNegative @LTLengthOf(value = "#2", offset = "#3 - 1") int len) throws IOException {
     checkNotNull(in);
     checkNotNull(b);
     if (len < 0) {
       throw new IndexOutOfBoundsException("len is negative");
     }
     int total = 0;
     while (total < len) {
+      @SuppressWarnings("argument.type.incompatible") /* if the offset and length passed to this method are invalid,
```

Comment:
(Posting again, because previous conversation got hidden)

Actually, to slightly negate what I said in our meeting, it seems that the code does not always work according to the specification. If len is 0, and offset is invalid, then it just returns 0, but the comment says it should throw. Not sure if it is a bug in implementation or bad wording in the comment, but maybe the checks really should be there.